### ISSUE TITLE
Perpetual output using EventFlow + ApplicationInsights + NLog

### ISSUE THREAD

**Issue by ibebbs** [Azure/diagnostics-eventflow#140]
## Description

I wish to output events written to an [EventSource](https://msdn.microsoft.com/en-us/library/system.diagnostics.tracing.eventsource.aspx) to both StdOut and ApplicationInsights. As such I used EventFlow to listen to events from the EventSource and write the output to both destinations (I had previously tried using [ApplicationInsights.EventSourceListener](https://www.nuget.org/packages/Microsoft.ApplicationInsights.EventSourceListener/) but, for reasons yet to be determined, it simply would not capture any output to ApplicationInsights). Anyway, this seemed to work fine and output is indeed written to both destinations as expected.

Now I also wish to add NLog output to ApplicationInsights so installed the [ApplicationInsights.NLogTarget](https://www.nuget.org/packages/Microsoft.ApplicationInsights.NLogTarget/). Again, this seemed to work as expected and messages written to NLog were indeed sent to ApplicationInsights.

Unfortunately, in the above scenario, if you delay writing to NLog long enough for libraries to initialize themselves fully, when the first NLog message is written you will see an unending stream of data containing event output from the "Microsoft-ApplicationInsights-Data" event source and find that the program enters a perpetual loop of writing data to the EventFlow output (including - most damagingly - the ApplicationInsights output).

## Reproduction

A full reproduction of the issue, with steps, can be found in this [repo](https://github.com/ibebbs/AiNlogEventSource)

## Investigation

By cloning EventFlow and building a debug version of Microsoft.Diagnostics.EventFlow.Inputs.EventSource (and temporarily disabling the strong name check), I was able to break at the pertinent points in the EventSourceInput.cs file. The following was noted:

1. A call to ```OnEventSourceCreated``` was made for the 'Microsoft-ApplicationInsights-Data' event source and passed to the 'EnableAsNecessary' method. As the 'Microsoft-ApplicationInsights-Data' was not included in the EventFlow config (and not a 'special case') no call to ```EnableEvents``` was made.
2. A subsequent call to ```OnEventWritten``` is (incorrectly) made for an event from the 'Microsoft-ApplicationInsights-Data' event source and, as it's not a 'special case', passed to the output subject and, ultimately, the output modules.

This can be seen in the call stack here:

![Call stack](https://raw.githubusercontent.com/ibebbs/AiNlogEventSource/master/Images/Call%20Stack.png)

This sequence of events ceases to occur in any of the following conditions:

1. The EventSource input for EventFlow is removed from configuration (or it's source list left empty due to the EventSourceInput configuration being invalid).
2. The ApplicationInsights output for EventFlow being removed from configuration
3. The ```<InstrumentationKey>``` element being removed from the ApplicationInsights.config file.
4. Output written to the NLog output synchronously at startup

The last point here is particularly pertinent. As can be seen in the code within _Program.cs_ (shown in the repo), a delay is introduced prior to writing to nlog to allow other libraries to complete initialization. __It is my belief that one of these libraries (perhaps Microsoft.Diagnostics.EventFlow.Outputs.ApplicationInsights) is enabling the 'Microsoft-ApplicationInsights-Data' EventSource causing the additional output and resultant recursive loop.__

I would very much appreciate help with this issue to understand why I am seeing output from the 'Microsoft-ApplicationInsights-Data' event source in the output stream. I believe this to be (a rather serious) bug in EventFlow (when mixed with the other packages above)

**Comment by karolz-ms** [comment]
Ian, thank you for a detailed issue report, much appreciated!

Let me see what I can find out...

**Comment by karolz-ms** [comment]
@ibebbs thanks again for a good repro--I was able to replicate the problem in house.

The Microsoft-ApplicationInsights-Data is being enabled by Application Insights SDK. What I do not understand is why events from that source end up being captured by EventFlow's EventSource input, since the input, as you observed, _is not_ enabling them.

I am still trying to get to the bottom of this.

**Comment by karolz-ms** [comment]
So it turns out to be a problem in the .NET Framework https://github.com/dotnet/coreclr/issues/14434 :-(

I think we need to have a workaround for this in EventFlow and in ApplicationInsights EventSource telemetry module, because it might take a while to have this problem fixed in the framework. I suggest we add a new setting that allows to explicitly "disable" some EventSources. It could be a regular expression that we would match against provider name, for example

```json
"disabledSourcesNamePattern": "Microsoft-ApplicationInsights*"
```

And while we are at this, we could also introduce `enabledSourceNamePattern` that would allow to enable several EventSources with a single setting.

@ibebbs @brahmnes @SergeyKanzhelev what do you guys think?

**Comment by SergeyKanzhelev** [comment]
Love it. Will you send PR?

**Comment by brahmnes** [comment]
Looks like a good capability to have. Instead of full blown regular expression, we can keep it simple, like the '*' is supported.

**Comment by karolz-ms** [comment]
@SergeyKanzhelev sure, ETA sometime next week

@brahmnes Regular expressions are well-supported in the framework, so there is no reason not to use them really.

**Comment by karolz-ms** [comment]
@brahmnes after thinking about it some more it is probably too much of a "good" thing to use regular expressions here. Instead, a list of enabled/disabled EventSource name prefixes should be both sufficient and much faster to process

**Comment by karolz-ms** [comment]
I just pushed EventSource input v 1.2.0 to Nuget.org. @ibebbs please take a look when you have a chance and see if you can make your application work by means of the new configuration option `disabledProviderNamePrefix`

I am going to keep this issue open to track the work needed for Application Insights EventSourceTelemetryModule

**Comment by ibebbs** [comment]
Thanks @karolz-ms.

I am only now coming back to look at this issue again but my plan is to try getting ```ApplicationInsights.EventSourceListener``` working before resorting to the ```disabledProviderNamePrefix```. This is mostly due to the merged change feeling a bit... ahem... hacky and it not being safe by default (i.e. if I forget to add the configuration - or someone else removes it for some reason - I will start emitting a tonne of data to ApplicationInsights - _again_!).

Once the root of the issue has been resolved I will certainly be looking to move back to EventFlow.

**Comment by karolz-ms** [comment]
OK, @ibebbs it would be extremely helpful if you let us know here how the ApplicationInsights.EventSourceListener works for you. I am familiar with that codebase and _it likely has the same problem_; in fact, one of my next work items is to add the same capability to AI.EventSourceListener that has just been added to EventFlow EventSource input.

**Comment by ibebbs** [comment]
Hi Karolz,

I've finally had an opportunity to look back at the ApplicationInsights.EventSourceListener package and have managed to get it to capture events. The issue outlined here does not _seem_ to effect this package (I have NLog  & EventSources writing to AI and am not seeing any events from 'Microsoft-ApplicationInsights-Data') so will be looking to move forward with this approach for now (with a custom processor for writing traces to the StdOut).

One issue here is that, unlike EventFlow.Inputs.EventSource, ApplicationInsights.EventSourceListener does not include the EventSource.ProviderName as a property on the output trace event. I am about to prepare and commit a PR for adding this property but, as you seem to be responsible for the ApplicationInsights-dotnet-logging repository too, I was wondering if you knew why this property was not being collected and whether there would be an issue merging a PR to include it.

Thanks,
  Ian

**Comment by karolz-ms** [comment]
Thanks, Ian, this is good to know. Looks like we have some breathing space regarding adding the workaround to ApplicationInsights.EventSourceListener. I still want to do that eventually, because I do think someone will run into this issue with AI.EventSourceListener eventually. But looks like you are not blocked, which is great to hear.

**Comment by karolz-ms** [comment]
The workaround for the EventListener issue is on track to be shipped with Application Insights SDK version 2.5.
